### PR TITLE
Patreon api

### PR DESCRIPTION
**Pull request by rubenvanassche**

### CODE REVIEW

File: .env.example [hunk 1]
```diff
@@ -41,4 +41,7 @@ BASIC_AUTH_PASSWORD=secret
 
 RESPONSE_CACHE_ENABLED=false
 
-INSTAGRAM_TOKEN=
\ No newline at end of file
+INSTAGRAM_TOKEN=
+
+PATREON_CLIENT_ID=
+PATREON_SECRET=
```
Comment (freekmurze) [inline .env.example]: Newline missing

File: app/Console/Commands/ImportPatreonPledgers.php [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+class ImportPatreonPledgers extends Command
+{
+    protected $signature = 'import:patreon-pledgers';
+
+    protected $description = 'Import pledgers from Patreon';
+
+    /** @var \App\Services\Patreon\PatreonApi */
+    protected $api;
+
```
Comment (freekmurze) [inline app/Console/Commands/ImportPatreonPledgers.php]: Remove this empty line

File: app/Console/Commands/ImportPatreonPledgers.php [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        $this->api = $api;
+
+        parent::__construct();
+    }
+
+    public function handle()
+    {
+        $this->info('Importing pledgers from Patreon...');
+
+        $campagin = $this->api->campaigns()->first();
```
Comment (freekmurze) [inline app/Console/Commands/ImportPatreonPledgers.php]: Fix the typo in variable name.

Comment (freekmurze) [reply app/Console/Commands/ImportPatreonPledgers.php]: Please handle the case for when no campaign is found.

File: app/Console/Commands/ImportPatreonPledgers.php [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+        $validRewards = $campagin->rewards->filter(function (Reward $reward) {
```
Comment (freekmurze) [inline app/Console/Commands/ImportPatreonPledgers.php]: Reward under the specified amount are also "valid", so let's renamed the variable to something more correct. Maybe let's just use `$rewards`.

File: app/Console/Commands/ImportPatreonPledgers.php [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+
+    public function __construct(PatreonApi $api)
+    {
+        $this->api = $api;
+
+        parent::__construct();
+    }
+
+    public function handle()
```
Comment (freekmurze) [inline app/Console/Commands/ImportPatreonPledgers.php]: This method body is somewhat long. Let's make it shorter by moving some logic to a method called `getPledges`

File: app/Console/Commands/ImportPatreonPledgers.php [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+use Illuminate\Console\Command;
+
+class ImportPatreonPledgers extends Command
+{
+    protected $signature = 'import:patreon-pledgers';
+
+    protected $description = 'Import pledgers from Patreon';
+
+    /** @var \App\Services\Patreon\PatreonApi */
+    protected $api;
```
Comment (freekmurze) [inline app/Console/Commands/ImportPatreonPledgers.php]: `api` is a too generic variable name. Let's rename it variable to `patreon`

File: app/Console/Commands/ImportPatreonPledgers.php [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+            return $reward->amount >= 5000;
+        });
+
+        $pledges = $this->api->pledges($campagin->id)->filter(function (Pledge $pledge) use ($validRewards) {
+            return $validRewards->contains(function (Reward $reward) use ($pledge) {
+                return $reward->id === $pledge->rewardId;
+            });
+        });
+
+        foreach ($pledges as $pledge) {
```
Comment (freekmurze) [inline app/Console/Commands/ImportPatreonPledgers.php]: Use a collection and a filter operation instead of a loop.

File: app/Models/Patreon.php [hunk 8]
```diff
@@ -0,0 +1,49 @@
+<?php
+
+namespace App\Models;
+
+use App\Services\Patreon\Resources\User;
+use Illuminate\Database\Eloquent\Model;
+use Spatie\MediaLibrary\HasMedia\HasMedia;
+use Spatie\MediaLibrary\HasMedia\HasMediaTrait;
+
+class Patreon extends Model implements HasMedia
```
Comment (freekmurze) [inline app/Models/Patreon.php]: `Patreon` is named to generic as there are multiple concepts in `Patreon`. Let's rename this to `PatreonPleder`.

File: app/Services/Patreon/PatreonApi.php [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+use App\Services\Patreon\Resources\Campaign;
+use App\Services\Patreon\Resources\Pledge;
+use App\Services\Patreon\Resources\ResourceCollection;
+use App\Services\Patreon\Resources\Reward;
+use App\Services\Patreon\Resources\User;
+use GuzzleHttp\Client;
+use Illuminate\Support\Collection;
+
+class PatreonApi
```
Comment (freekmurze) [inline app/Services/Patreon/PatreonApi.php]: Let's rename this to `Patreon`

File: app/Services/Patreon/PatreonApi.php [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+{
+    /** @var \GuzzleHttp\Client */
+    protected $client;
+
+    public function __construct(string $accessToken)
```
Comment (freekmurze) [inline app/Services/Patreon/PatreonApi.php]: Accept a `Client` as a constructor argument. Move the setup of that client to a serviceprovider.

File: app/Services/Patreon/PatreonApi.php [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+            'base_uri' => 'https://www.patreon.com/api/oauth2/api/',
+            'headers' => [
+                'authorization' => "Bearer {$accessToken}",
+            ],
+        ]);
+    }
+
+    protected function request(string $endpoint): array
+    {
+        $response = null;
```
Comment (freekmurze) [inline app/Services/Patreon/PatreonApi.php]: Can be removed.

File: app/Services/Patreon/Resources/Campaign.php [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+    public $id;
+
+    /** @var string */
+    public $name;
+
+    /** @var int */
+    public $pledgedSum;
+
+    /** @var \App\Services\Patreon\Resources\ResourceCollection  */
+    public $rewards;
```
Comment (freekmurze) [inline app/Services/Patreon/Resources/Campaign.php]: Do we need to display this somewhere? If not, remove it. For all these resources only keep what you actually need.

File: app/Services/Patreon/Resources/Campaign.php [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+    public function __construct(int $id, string $name, int $pledgedSum)
+    {
+        $this->id = $id;
+        $this->name = $name;
+        $this->pledgedSum = $pledgedSum;
+
+        $this->rewards = new ResourceCollection();
+    }
+
+    public function fillRewards(array $rewardRelationsData, ResourceCollection $rewards)
```
Comment (freekmurze) [inline app/Services/Patreon/Resources/Campaign.php]: I don't think we are going to display rewards. Remove all code regarding rewards.

File: app/Services/Patreon/Resources/Resource.php [hunk 14]
```diff
@@ -0,0 +1,13 @@
+<?php
+
+namespace App\Services\Patreon\Resources;
+
+use phpDocumentor\Reflection\Types\Boolean;
+
+class Resource
```
Comment (freekmurze) [inline app/Services/Patreon/Resources/Resource.php]: Do we need this class?

File: app/Services/Patreon/Resources/Campaign.php (hunk 12 above)
Comment (rubenvanassche) [reply app/Services/Patreon/Resources/Campaign.php]: They are being used when filtering out the rewards with an amount above €50

```php
        $rewards = $campaign->rewards->filter(function (Reward $reward) {
            return $reward->amount >= 5000;
        });
```

File: database/migrations/2018_10_04_165241_create_patreon_pledgers_table.php [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+        Schema::create('patreon_pledgers', function (Blueprint $table) {
+            $table->increments('id');
+            $table->integer('patreon_id');
+            $table->string('name');
+            $table->string('url_to_original');
+            $table->timestamps();
+        });
+    }
+
+    /**
```
Comment (freekmurze) [inline database/migrations/2018_10_04_165241_create_patreon_pledgers_table.php]: We never use down migrations, you can remove this

File: database/migrations/2018_10_04_165241_create_patreon_pledgers_table.php [hunk 16]
```diff
@@ -0,0 +1,34 @@
+<?php
+
+use Illuminate\Database\Migrations\Migration;
+use Illuminate\Database\Schema\Blueprint;
+use Illuminate\Support\Facades\Schema;
+
+class CreatePatreonPledgersTable extends Migration
+{
+    /**
```
Comment (freekmurze) [inline database/migrations/2018_10_04_165241_create_patreon_pledgers_table.php]: Remove this useless comment

File: database/seeds/PatreonPledgersSeeder.php [hunk 17]
```diff
@@ -0,0 +1,18 @@
...
+
+use Illuminate\Database\Seeder;
+
+class PatreonPledgersSeeder extends Seeder
+{
+    public function run()
+    {
+        factory(\App\Models\PatreonPledger::class, 10)
+            ->create()
+            ->each(function (\App\Models\PatreonPledger $patreon) {
```
Comment (freekmurze) [inline database/seeds/PatreonPledgersSeeder.php]: Import that namespace

**Pull request merged**